csharp/src/Google.Protobuf/ExtensionSet.cs

```diff
@@ -0,0 +1,341 @@
+#endregion
+
+using Google.Protobuf.Collections;
+using System;
+using System.Collections.Generic;
+using System.Linq;
+
+namespace Google.Protobuf
+{
+    /// <summary>
+    /// Methods for managing <see cref="ExtensionSet{TTarget}"/>s with null checking.
+    /// 
+    /// Most users will not use this class directly
+    /// </summary>
+    public static class ExtensionSet
+    {
+        private static bool GetValue<TTarget>(ref ExtensionSet<TTarget> set, Extension extension, out IExtensionValue value) where TTarget : IExtensionMessage<TTarget>
+        {
+            if (set == null)
+            {
+                value = null;
+                return false;
+            }
+            return set.ValuesByNumber.TryGetValue(extension.FieldNumber, out value);
+        }
+
+        /// <summary>
+        /// Registers the specified extension in this set
+        /// </summary>
+        public static void Register<TTarget>(ref ExtensionSet<TTarget> set, Extension extension) where TTarget : IExtensionMessage<TTarget>
+        {
+            if (set == null)
+            {
+                set = new ExtensionSet<TTarget>();
+            }
+            if (extension.TargetType != typeof(TTarget))
+            {
+                throw new ArgumentException("Cannot register extension for wrong target type");
+            }
+            if (set.ValuesByNumber.ContainsKey(extension.FieldNumber))
+            {
+                throw new ArgumentException("Set already contains an extension with the specified field number");
+            }
+            set.ValuesByNumber.Add(extension.FieldNumber, extension.CreateValue());
+        }
+
+        private static TValue GetDefaultValue<TValue>() 
+        {
+            if (typeof(TValue) == typeof(string)) 
+            {
+                return (TValue)(object)"";
+            }
+            else if (typeof(TValue) == typeof(ByteString))
+            {
+                return (TValue)(object)ByteString.Empty;
+            }
+            else
+            {
+                return default(TValue);
+            }
+        }
+
+        /// <summary>
+        /// Gets the value of the specified extension
+        /// </summary>
+        public static TValue Get<TTarget, TValue>(ref ExtensionSet<TTarget> set, Extension<TTarget, TValue> extension) where TTarget : IExtensionMessage<TTarget>
+        {
+            IExtensionValue value;
+            if (GetValue(ref set, extension, out value))
+            {
+                return ((ExtensionValue<TValue>)value).GetValue();
+            }
+            else 
+            {
+                return GetDefaultValue<TValue>();
+            }
+        }
+
+        /// <summary>
+        /// Gets the value of the specified repeated extension
+        /// </summary>
+        public static RepeatedField<TValue> Get<TTarget, TValue>(ref ExtensionSet<TTarget> set, RepeatedExtension<TTarget, TValue> extension) where TTarget : IExtensionMessage<TTarget>

```

Comment:
I'm actually unsure about this one. I have a couple ways I could implement it and they each have pros and cons but I'm unsure which would be the best fit for the lib.

1. Current implementation
Pros: Matches with other repeated field APIs. Doesn't require null checks or separate "RegisterExtension" calls.
Cons: Simple calls to read members such as Length can allocate a extension set, extension value, and repeated field.

2. Return null on empty set or non-registered value
Pros: Calls to check properties like length don't allocate
Cons: Null checks required which don't match other APIs that use repeated fields. RegisterExtension call required to create and use field

3. Match the Java implementation
Pros: Doesn't require null checks and calls to length properties don't allocate.
Cons: Massive difference in API. I personally wouldn't chose it since it's not very "Google.Protobuf" in its API.

Reply:
Another alternative I've just considered is a `GetOrRegister` extension method. Get returns null as to not allocate and the extension method will register if Get returns null, thus not requiring a separate call.

Reply:
For proposal 2, how users set/add repeated field if return null?

Reply:
A call to RegisterExtension would be required to create and use the field

Reply:
Both 1 and 2 make sense to me. I prefer 1

Reply:
I'm actually starting to prefer option two with the "GetOrRegister" method, given how gRPC is getting integrated with ASP.NET Core and with their focus on low memory and speed, it'd make sense to try and keep allocations low.